### PR TITLE
Plugin Directory: Stop block validator falsely warning about missing script translations

### PR DESCRIPTION
**Pull request by obenland**
## Summary

The block plugin validator at https://wordpress.org/plugins/developers/block-plugin-validator/ shows the warning **"No translations are loaded for the scripts."** for plugins that use the modern `register_block_type( __DIR__ )` pattern with a `textdomain` declared in `block.json`. Since WordPress 5.7, core automatically calls `wp_set_script_translations()` for every script handle declared in `block.json` (`script`, `viewScript`, `editorScript`, etc.) when a `textdomain` is set, so the warning is a false positive in that case.

`Block_Plugin_Checker::check_for_translation_function()` only searched the parsed PHP function calls for a literal `wp_set_script_translations` and warned otherwise, regardless of whether translations were actually being registered.

## Changes

- `cli/class-block-plugin-checker.php` — `check_for_translation_function()` now also suppresses the warning when the plugin calls `register_block_type` / `register_block_type_from_metadata` **and** at least one parsed `block.json` declares a `textdomain`. If `register_block_type` is called but no `block.json` sets `textdomain`, the warning still fires (translations genuinely won't load in that case).
- `shortcodes/class-block-validator.php` — updated the detailed help text for that check to mention the auto-registration path so authors aren't told to add a redundant manual call.

## Test plan

- [ ] On `/developers/block-plugin-validator/`, run the checker against a block plugin that registers via `register_block_type( __DIR__ )` with `"textdomain": "..."` in `block.json` and no explicit `wp_set_script_translations()` — confirm the "No translations are loaded for the scripts." warning is gone.
- [ ] Run the checker against a plugin that calls `register_block_type` without a `textdomain` in `block.json` — the warning should still appear.
- [ ] Run the checker against a plugin that calls `wp_set_script_translations()` directly — no warning (existing behavior, unchanged).
- [ ] Open the warning's `<details>` on a plugin that does trigger it, and verify the updated help text reads correctly.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 1]
```diff
@@ -1019,18 +1019,39 @@ function check_total_size() {
...
-			);
+
+		if ( in_array( 'wp_set_script_translations', $functions, true ) ) {
+			return;
 		}
+
+		$registers_block_from_metadata = (
+			in_array( 'register_block_type', $functions, true ) ||
+			in_array( 'register_block_type_from_metadata', $functions, true )
+		);
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php]: Fixed in 2daae16a5 — switched `find_php_functions()` from `$all_calls += $php_calls` to `array_push( $all_calls, ...$php_calls )` so per-file results are appended rather than unioned. Each file's calls are numerically indexed from 0, so the union was dropping every entry whose key collided with an earlier file. Good catch — without this, the new translation check could miss `register_block_type()` in any file beyond the first.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 2]
```diff
@@ -1019,18 +1019,39 @@ function check_total_size() {
...
+		$has_block_json_textdomain = false;
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php]: Fixed in 2daae16a5 — renamed to `$registers_block_type` since the variable also covers the classic `register_block_type( 'namespace/name', $args )` form.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/shortcodes/class-block-validator.php [hunk 3]
```diff
@@ -435,7 +435,7 @@ public static function get_detailed_help( $method, $result ) {
 			case 'check_for_translation_function':
 				return sprintf(
 					// translators: %s is the link to the internationalization docs.
-					__( 'Block plugins should use <code>wp_set_script_translations</code> to load translations for each script file. <a href="%s">Learn more about internationalization.</a>', 'wporg-plugins' ),
+					__( 'Block plugins should load translations for each script file, either by declaring a <code>textdomain</code> in <code>block.json</code> and registering the block via <code>register_block_type()</code> (which calls <code>wp_set_script_translations()</code> automatically since WordPress 5.7), or by calling <code>wp_set_script_translations()</code> directly. <a href="%s">Learn more about internationalization.</a>', 'wporg-plugins' ),
 					'https://developer.wordpress.org/block-editor/developers/internationalization/'
 				);
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/shortcodes/class-block-validator.php:440]: Fixed in 2daae16a5 — help text now mentions both `register_block_type()` and `register_block_type_from_metadata()` so authors using either are covered.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 4]
```diff
@@ -1019,18 +1023,39 @@ function check_total_size() {
...
+		$registers_block_type = (
+			in_array( 'register_block_type', $functions, true ) ||
+			in_array( 'register_block_type_from_metadata', $functions, true )
+		);
+		$has_block_json_textdomain = false;
+		foreach ( (array) $this->blocks as $block ) {
+			if ( ! empty( $block->textdomain ) ) {
+				$has_block_json_textdomain = true;
+				break;
+			}
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php]: Fixed in a7139c414 — detection now source-scans for the metadata-based registration pattern (`register_block_type_from_metadata(`, or `register_block_type(` whose first argument is anything other than a `'ns/name'` string literal). The classic name form alone no longer suppresses the warning. Result is cached on the instance so the unit tests can seed it via reflection.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/shortcodes/class-block-validator.php [hunk 5]
```diff
@@ -435,7 +435,7 @@ public static function get_detailed_help( $method, $result ) {
 			case 'check_for_translation_function':
 				return sprintf(
 					// translators: %s is the link to the internationalization docs.
-					__( 'Block plugins should use <code>wp_set_script_translations</code> to load translations for each script file. <a href="%s">Learn more about internationalization.</a>', 'wporg-plugins' ),
+					__( 'Block plugins should load translations for each script file, either by declaring a <code>textdomain</code> in <code>block.json</code> and registering the block via <code>register_block_type()</code> or <code>register_block_type_from_metadata()</code> (which calls <code>wp_set_script_translations()</code> automatically since WordPress 5.7), or by calling <code>wp_set_script_translations()</code> directly. <a href="%s">Learn more about internationalization.</a>', 'wporg-plugins' ),
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/shortcodes/class-block-validator.php]: Fixed in a7139c414 — help text now reads "Registering your block from its `block.json` metadata (e.g. `register_block_type( __DIR__ )` or `register_block_type_from_metadata()`)…" and explicitly contrasts that with calling `wp_set_script_translations()` directly.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 6]
```diff
@@ -1019,18 +1021,88 @@ function check_total_size() {
...
+			 * literal as its first argument is metadata-based — the argument is a path,
+			 * __DIR__, dirname() expression, etc. The classic name form does not
+			 * trigger core's automatic wp_set_script_translations() wiring.
+			 *
+			 * `\s*+` is possessive so the engine can't backtrack into the whitespace
+			 * and bypass the negative lookahead.
+			 */
+			if ( preg_match( '#\bregister_block_type\s*\(\s*+(?![\'"][\w-]+/[\w-]+[\'"]\s*[,)])#', $contents ) ) {
+				$this->registers_block_from_metadata = true;
+				break;
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php]: Fixed in 024d1af0e — replaced the raw-source regex with a PHP tokenizer pass (`token_get_all` + `TOKEN_PARSE`). Only real `T_STRING` calls to `register_block_type` are inspected; the first non-whitespace token after `(` is the first argument. Comments (`T_COMMENT` / `T_DOC_COMMENT`) and string literals (`T_CONSTANT_ENCAPSED_STRING`) containing `register_block_type( __DIR__ )` text no longer influence the result. Verified against 12 cases including misleading comments and string literals.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+			array(
+				(object) array(
+					'name'       => 'plugin/main',
+					'textdomain' => 'plugin',
+				),
+			),
+			false
+		);
+		$this->assertCount( 1, $warnings );
+	}
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php:112]: Fixed in 024d1af0e — consolidated the two duplicate cases into `test_no_metadata_registration_warns_even_with_textdomain`, with a docblock that's explicit about what the test actually exercises (metadata-helper false + textdomain present, regardless of the source of the false verdict).

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 8]
```diff
@@ -430,7 +431,8 @@ public function find_php_functions( $base_dir ) {
 					[ $filename, $php_calls->get_error_data() ]
 				);
 			} else {
-				$all_calls += $php_calls;
+				// Append, not union — per-file calls share numeric keys from 0 so `+=` would drop later files.
+				array_push( $all_calls, ...$php_calls );
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php:434]: Not applied — verified that `array_push( $arr, ...[] )` is a no-op since PHP 7.3 (`array_push` accepts a single-array call and the empty spread degrades to exactly that). No warning is raised. `array_merge` would also work but the existing form is correct.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 9]
```diff
@@ -1019,18 +1021,149 @@ function check_total_size() {
...
+			}
+			if ( $j >= $count ) {
+				continue;
+			}
+
+			$first_arg = $tokens[ $j ];
+			if ( ! is_array( $first_arg ) || T_CONSTANT_ENCAPSED_STRING !== $first_arg[0] ) {
+				// Variable, __DIR__, dirname() call, expression, etc. — metadata-based.
+				return true;
+			}
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php]: Not applied — the failure mode (a plugin defining its block name as a PHP global constant and calling `register_block_type( MY_BLOCK, $args )`) is exotic enough that I don't think it justifies expanding the helper. Tightening it to only treat known path-shaped tokens as metadata would also reject the legitimate `register_block_type( $path )` pattern, swapping one rare edge case for another (and arguably a more common one). Leaving the trade-off as-is for now; if a real plugin trips this we can revisit.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php [hunk 10]
```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * Tests for Block_Plugin_Checker::check_for_translation_function().
+ *
+ * @package WordPressdotorg\Plugin_Directory\Tests
+ */
+
+use PHPUnit\Framework\Attributes\CoversMethod;
+use PHPUnit\Framework\TestCase;
+use WordPressdotorg\Plugin_Directory\CLI\Block_Plugin_Checker;
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php:10]: Not applied — this repo's plugin-directory test suite is on PHPUnit 11 (PR #588 moved it). Verified `tests-cli` reports `PHPUnit 11.1.0`, and `.github/workflows/unit-tests.yml` pins `phpunit:^11` for standalone jobs. `PHPUnit\Framework\Attributes\CoversMethod` is supported.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+/**
+ * Verifies the rules under which the block validator's translation check fires.
+ *
+ * @group block-validator
+ */
+#[CoversMethod( Block_Plugin_Checker::class, 'check_for_translation_function' )]
+class Block_Plugin_Checker_Translation_Test extends TestCase {
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php:18]: Not applied — same reason as the import-line thread: the plugin-directory suite runs on PHPUnit 11.1.0 (verified in `tests-cli`), so the attribute resolves and `@covers` would be the deprecated form here.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+		$blocks_prop = $reflection->getProperty( 'blocks' );
+		$blocks_prop->setAccessible( true );
+		$blocks_prop->setValue( $checker, $blocks );
+
+		$cache_prop = $reflection->getProperty( 'registers_block_from_metadata' );
+		$cache_prop->setAccessible( true );
+		$cache_prop->setValue( $checker, $registers_block_from_metadata );
+
+		$checker->check_for_translation_function();
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php]: Valid concern — these tests go through the cache and don't exercise the tokenizer pass in `file_has_metadata_register_block_type_call()`. The new logic was verified informally against 12 representative cases (comments, string literals, classic name, path expressions, mixed) but those checks aren't in the suite. Holding off on adding them here to keep this PR's scope under control; if the tokenizer-based detection sticks I'll add the coverage as a follow-up.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 13]
```diff
@@ -1019,18 +1021,149 @@ function check_total_size() {
...
+			while ( $j < $count && is_array( $tokens[ $j ] ) && T_WHITESPACE === $tokens[ $j ][0] ) {
+				$j++;
+			}
+			if ( $j >= $count || '(' !== $tokens[ $j ] ) {
+				continue;
+			}
+
+			// First non-whitespace token after `(` is the first argument.
+			$j++;
+			while ( $j < $count && is_array( $tokens[ $j ] ) && T_WHITESPACE === $tokens[ $j ][0] ) {
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php]: Fixed in 99714b8a2 — the first-argument scan now also skips `T_COMMENT` and `T_DOC_COMMENT` tokens, so `register_block_type( /* name */ 'ns/block', $args )` correctly resolves to the classic name form. Verified against 11 representative cases including inline comment + classic, inline comment + metadata, and docblock + classic.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 14]
```diff
@@ -1019,18 +1021,149 @@ function check_total_size() {
...
+			return false;
+		}
+
+		$count = count( $tokens );
+		for ( $i = 0; $i < $count; $i++ ) {
+			$token = $tokens[ $i ];
+			if ( ! is_array( $token ) || T_STRING !== $token[0] || 'register_block_type' !== $token[1] ) {
+				continue;
+			}
+
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php]: Fixed in 99714b8a2 — the helper now checks the previous significant token (skipping whitespace and comments) and ignores `register_block_type` when it follows `T_OBJECT_OPERATOR`, `T_NULLSAFE_OBJECT_OPERATOR`, `T_DOUBLE_COLON`, `T_FUNCTION`, or `T_NEW`. So `$obj->register_block_type(...)`, `Foo::register_block_type(...)`, `$obj?->register_block_type(...)`, and `function register_block_type(...)` declarations no longer count toward metadata-registration.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php [hunk 15]
```diff
@@ -1019,18 +1020,34 @@ function check_total_size() {
...
+		$registers_block = in_array( 'register_block_type', $functions, true )
+			|| in_array( 'register_block_type_from_metadata', $functions, true );
+
+		if ( $registers_block ) {
+			foreach ( (array) $this->blocks as $block ) {
+				if ( ! empty( $block->textdomain ) ) {
+					return;
+				}
+			}
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php:1042]: Not applied — this was a deliberate design choice. Earlier commits on this branch (a7139c414 through c56f0654c) did exactly the tokenizer-based disambiguation you're describing, and we removed it in 5fddcb7f0 because the scaffolding (~130 lines) was disproportionate to the corner case it caught: a plugin that uses classic-form `register_block_type( 'ns/name', $args )` AND has a `textdomain` declared in `block.json`. In practice, `textdomain` only ends up in `block.json` if the plugin is going through the metadata-registration code path, so the simpler "any `register_block_type*` call + any block.json textdomain → assume auto-loaded" rule has very low false-negative risk. The verbose docblock that previously called this out has been trimmed to match.

File: wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+	 * Block registration without any block.json textdomain still warns — core won't
+	 * auto-register translations.
+	 */
+	public function test_block_registration_without_textdomain_warns() {
+		$warnings = $this->run_check(
+			array( 'register_block_type_from_metadata' ),
+			array( (object) array( 'name' => 'plugin/main' ) )
+		);
+		$this->assertCount( 1, $warnings );
+	}
```
[the remark this answers is not in the document]
Comment (obenland) [reply wordpress.org/public_html/wp-content/plugins/plugin-directory/tests/Block_Plugin_Checker_Translation_Test.php:88]: Partially applied — added a fixture-based test (`test_fixture_register_block_type_with_dir_passes`) covering `register_block_type( __DIR__ )` + `block.json` with a `textdomain` in b1ed61a5d. The negative case for classic-form `register_block_type( 'ns/name', $args )` was deliberately not added because the simplified check (see sibling thread) does not distinguish classic-form from metadata-form — under the chosen design, classic-form + a block.json textdomain passes the check by design. Asserting the opposite would lock in behavior we explicitly chose not to enforce.
